docs/internal.rst

```diff
@@ -6,6 +6,68 @@ Intermediate representation
 Use ``ti.init(print_ir=True)`` to print IR on the console.
 
 
+Data Structure Node Organization
+--------------------------------
+
+The internal organization of data structure nodes can be confusing.
+It is important to distinguish the concept of **containers**, **cells** and **components**.
+
+- A **container** can have multiple **cells**. The numbers of **cells** are recommended to be powers of two.
+- A **cell** can have multiple **components**.
+- Each **component** is a **container** of a lower-level SNode.
```

Comment:
Could you say:
```suggestion
1. A **container** have multiple **cells**.

2. A **cell** have multiple **components**.

For example:

.. code-block :: python

    blk1 = ti.root                    # blk1 is a cell
    blk1.place(x, y)                 # x, y are components of blk1
    blk2 = blk1.dense(ti.i, 3)  # blk2 is a compoment of blk1

3. Each **component** is a **container** of a lower-level SNode.

.. note ::

     The numbers of **cells** are recommended to be powers of two.
```
and add @k-ye's picture in his slides?

Reply:
Thanks for the inputs, but `blk1` is a container/SNode instead of a cell...

Reply:
I get more puzzled... so we have 4 concepts: SNode, cell, components, container. Could you shrink down to 2 concepts? I guess most of them are the same, like components = container.

Reply:
IIUC, component and container are two perspectives of the same thing.

* A cell can contain multiple components, because we can attach multiple children SNodes to the same parent SNode.
* Each SNode itself is a container of multiple cells.

If you haven't, I would strongly suggest to read on the Taichi paper, particularly Section 3.2. Once you follow through those examples, things should be much more obvious.

Reply:
Reopened? I'm still puzzled with the multiple terms here.
Can we add an image here? Like the one in legacy taichi.